src/Serializing/SExpression/Precise.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE DataKinds, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, OverloadedStrings, ScopedTypeVariables, TypeApplications, TypeFamilies, TypeOperators, UndecidableInstances #-}
+module Serializing.SExpression.Precise
+( serializeSExpression
+) where
+
+import Data.ByteString.Builder
+import Data.Foldable (fold)
+import Data.List (intersperse)
+import Data.Text (Text)
+import GHC.Generics
+
+serializeSExpression :: ToSExpression t => t -> Builder
+serializeSExpression t = toSExpression t 0 <> "\n"
+
+
+nl :: Int -> Builder
+nl n | n <= 0    = ""
+     | otherwise = "\n"
+
+pad :: Int -> Builder
+pad n = stringUtf8 (replicate (2 * n) ' ')
+
+
+class ToSExpression t where
+  toSExpression :: t -> Int -> Builder
+
+instance (ToSExpressionWithStrategy strategy t, strategy ~ ToSExpressionStrategy t) => ToSExpression t where
+  toSExpression = toSExpressionWithStrategy @strategy undefined
+
+
+data Strategy = Generic | Show
+
+type family ToSExpressionStrategy t :: Strategy where
+  ToSExpressionStrategy Text = 'Show
+  ToSExpressionStrategy _    = 'Generic
```

Comment:
Our old friend advanced overlap (cf https://wiki.haskell.org/GHC/AdvancedOverlap, Oleg Kiselyov, Simon Peyton Jones) allows us to specialize the instances s.t.:

- `Text` fields are `show`n,
- everything else is handled via the generic `GToSExpression` class, and
- we don’t require instances of `ToSExpressionWithStrategy` (below) for every possible AST type (which would be agonizing).

This means that specializations of the behaviour for specific AST types will have to be listed in here—which makes this specific approach suitable only for cases where we specifically do _not_ intend to specialize the behaviour for any particular language’s AST. That is, if we sometimes wanted to customize the behaviour, we should take a different (tho perhaps similar) approach where we _do_ explicitly define specialized `ToSExpressionWithStrategy` instances for each datatype in each AST.

It may be possible to do something clever e.g. with `-XDerivingVia` or type families that would allow you to customize the behaviour piecemeal without necessarily requiring per-AST-datatype instances, but this sufficed for this particular case so I haven’t explored it further.